### ISSUE TITLE
Select none in update

### ISSUE THREAD

**Issue by Sytten** [prisma/prisma#6252]
## Problem
Currently an `update` is quite slow because it does 3 steps:
1. Select my thing
2. Update my thing
3. Select my thing

I do understand the need for the first select to make sure it updates only one record, but the last one if a bit annoying.

## Suggested solution

It was suggested elsewhere that it could be using a `RETURNING`/other similar clause to remove the last one.
This would be nice, but my understanding is that it is an important change.
I suggest as a first step to allow users to tell prisma to not return anything.
This would need a new keyword in `select`, I suggest `none`:

```typescript
prisma.myThing.update({
  data: { ... },
  where: { id: 1 },
  select: { none: true }
})
```

## Alternatives

You can use an `updateMany` to avoid both `select`.

**Comment by pantharshit00** [comment]
Related ? https://github.com/prisma/prisma/issues/5043

**Comment by Sytten** [comment]
Yes related, but not the same request. Sometimes you don't need the last select nor a returning.

**Comment by AradAlvand** [comment]
Even when you set `select` to an empty object, like:
```typescript
prisma.course.update({
    where: { id: 1 },
    select: {}
});
```
It throws a runtime error as follows:
```
The `select` statement for type Purchase must not be empty. Available options are listed in green.
``
Which makes ZERO sense. I DON'T want any data to be selected after updating, simple as that.

**Comment by tekoh** [comment]
no updates on this? seems like its been an issue for a while now

**Comment by mrodz** [comment]
https://www.prisma.io/docs/concepts/components/prisma-client/null-and-undefined
☝️ The docs state:

> `null` is a value
> `undefined` means do nothing

So, you can achieve the behavior you're looking for by writing:
```js
prisma.myThing.update({
  data: { ... },
  where: { id: 1 },
  select: undefined
})
```

Hope this helps :)

**Comment by JonathanWilbur** [comment]
Does `select: null` produce these `SELECT` queries? It seems to be entirely undocumented what `select: null` does in mutating queries.

**Comment by tekoh** [comment]
> Does `select: null` produce these `SELECT` queries? It seems to be entirely undocumented what `select: null` does in mutating queries.

it returns all fields

**Comment by janpio** [comment]
(I think the mention of using a query with `RETURNING` in the original message here confused many people for a while, but I reformatted it now so it is clear that the suggestion is to just not try to get the results of the `UPDATE` at all.)

**Comment by paulintrognon** [comment]
Related https://github.com/prisma/prisma/issues/4246

**Comment by jeremyisatrecharm** [comment]
I also just ran into this and it is very confusing / wasteful

**Comment by JustMendex** [comment]
This really should be a thing especially if you are using a cloud based database paying per read and writes the reads in this case just  add up and keep adding up.

**Comment by Zenoo** [comment]
I'm hoping this gets implemented soon, I'm tired of having to add `{select: { id: true } }` to every update request.

**Comment by adri1wald** [comment]
worse than the fact you have to specify `select: { id: true }` is the fact that not doing so is not a type error, but a runtime error!

**Comment by markobalogh** [comment]
Adding myself to the mix as still hoping for this to get fixed!!

**Comment by paulintrognon** [comment]
> worse than the fact you have to specify `select: { id: true }` is the fact that not doing so is not a type error, but a runtime error!

What do you mean? How can you generate a runtime error by not adding `select: { id: true }`? If you don't specify the select, everything will be returned - including the id

**Comment by JonathanWilbur** [comment]
I think he means that `select: {}` wont return an error at compile time, but it will throw at runtime.

**Comment by muriloGervasio** [comment]
In adherence to the CQS pattern, data mutation operations ideally do not return values, and introducing an additional select query for confirmation may seem counterintuitive. To streamline this process, a potential workaround involves leveraging middlewares. By utilizing middlewares, we can automatically set the select field, reducing the number of fields and relations, minimizing the overall effort required for post-mutation actions.

**Comment by muriloGervasio** [comment]
```
this.$extends({
      name: 'middleware',
      query: {
        $allModels: {
          async update({ query, args }) {
            args.select = { id: true };
            return query(args);
          },
        },
      },
    });
```

**Comment by Fingertips18** [comment]
I would love to have this feature soon. Sometimes it's better to return nothing, especially if there is no reason to, and it's a plus for optimization as well.

**Comment by rolldeep-stepmerrily** [comment]
![images](https://github.com/prisma/prisma/assets/98504913/930e2008-8ed0-4ab0-8b83-8240d9185379)

I strongly agree with this opinion. I always include "select: { id: true }" in the update query, and I think adding this is very wasteful. If you don't add this select code, it will select all columns of the updated record, which wastes database read costs.

**Comment by CuriouslyCory** [comment]
Given the rise in popularity of managed databases billed on usage, this may become a critical issue for our team soon. With vercel's postgres we're billed on data transfer and it would be really nice to minimize the amount of data being returned by my upsert queries, especially when I'm not concerned with returning the values. 
Having an option for an "optimistic" return on upsert would be another great option where the values being passed into the create/update are used to populate the return unless there's an error.

**Comment by JonathanWilbur** [comment]
Man, this issue has been open for three years. How hard is it to _not_ do an extraneous query?

**Comment by toteto** [comment]
The same is usefull for other queries, like `findUnique`/`findFirst` where I want to simply check the existence of such entry but I am not interested in the values it might hold

```
const accountEmail = '...';
const mailInUse = await this.prisma.accounts.findUnique({ where: { email: accountEmail }, select: {} }) 
if (mailInUse) throw new Error(`Email ${accountEmail} is already used by existing account`);
// do whatever
```

**Comment by Zenoo** [comment]
> The same is usefull for other queries, like `findUnique`/`findFirst` where I want to simply check the existence of such entry but I am not interested in the values it might hold
> 
> ```ts
> const accountEmail = '...';
> const mailInUse = await this.prisma.accounts.findUnique({ where: { email: accountEmail }, select: {} }) 
> if (mailInUse) throw new Error(`Email ${accountEmail} is already used by existing account`);
> // do whatever
> ```

You can use `.count` instead of `.findFirst` for that.

**Comment by toteto** [comment]
> You can use `.count` instead of `.findFirst` for that.

Yes, but no, in case where there is existing entry, `.count` will search over the while table for matches and return the count (probably 1). `.findFirst`/`.findUnique` will exit the search at the first hit. Probably the performance would be about the same because the complexity is the same, but it would not hurt to have the option to do so.

**Comment by gastonyte** [comment]
I am very impatient to see this possibility, in the name of performance, it would be smart to allow it.

Who can do more, can do less

**Comment by MJomaa** [comment]
Just add a

```
select: {
  id: true // SELECT NONE
}

```
for now, so you can come back later and change it to `select: false` or `select {}` when it's implemented

**Comment by turnerEP** [comment]
It really grinds my gears that every time I come back to this issue it is still not really addressed. 
It seems so trivial to implement and given the overall great developer experience & features prisma offers I don't get why this issue is still a thing.
To me prisma really is a superb ORM with a great approach to avoid overfetching data without having type issues; the behavior of update / create is just odd in that context..

**Comment by Red-Killer** [comment]
+1

**Comment by WuDo** [comment]
+1

**Comment by s-kinch** [comment]
+1

**Comment by Winston-Hsiao** [comment]
+1

**Comment by adpopescu338** [comment]
+1

**Comment by b10902118** [comment]
+1

**Comment by wavim** [comment]
bump

**Comment by Toscanah** [comment]
+1

**Comment by n-hlyustin** [comment]
+1

**Comment by nikita-o** [comment]
+1

**Comment by drakedeatonuk** [comment]
+1

**Comment by DamianoMagrini** [comment]
This is a very relevant feature, but please remember that everyone who's subscribed to the issue gets an email when a comment is added. The :+1: button under the first message should be a good replacement to show interest! Daily bumps are definitely unnecessary.

**Comment by morrid0** [comment]
Another approach would be to have an `updateAndReturn`, as there's already an `updateManyAndReturn`.

**Comment by orlando-fv** [comment]
+1

**Comment by Th3Vladimir** [comment]
+1

**Comment by ellie-xpawn** [comment]
+1

**Comment by EvanSmith93** [comment]
+1

**Comment by charliehoover-jpg** [comment]
+1

**Comment by Yohannfra** [comment]
> I am quite surprised that this issue is getting so little attention from the community, as it is in my opinion quite a simple way of optimizing DB load and response time and [costs](https://github.com/prisma/prisma/issues/6252#issuecomment-2246650473).
> 
> Currently I feel like I should add `select: { id: true }` to my update queries so that they don't select all of my numerous fields for no reason, and it does not feel clean at all.

I did this thank

```ts
// In prisma we can't select nothing so this the best we can do for now 
// see this issue: https://github.com/prisma/prisma/issues/6252
export const PRISMA_EMPTY_SELECT_HACK = { id : true } as const
```

**Comment by lky141319** [comment]
Claiming bounty. 
Wallet: 0xd2268DEDca16b08aF4bb6Ad20cBa877D10cf8Ac6
GitHub: @lky141319

**Comment by lky141319** [comment]
Claiming bounty. 
Wallet: 0xd2268DEDca16b08aF4bb6Ad20cBa877D10cf8Ac6
GitHub: @lky141319

**Comment by sashaiwan** [comment]
I think I can take this one. this issue is open, but still needed?

**Comment by ogroppo** [comment]
> but still needed?

Deffo needed and useful :)

**Comment by rolki-png** [comment]
up
